src/alarms/alarm-handling.md

```diff
@@ -0,0 +1,87 @@
+
+| Name              | Description   |
+| ----------------- |---------------|
+| ref-string        | Device-unique reference string (kind of alarm ID or alarm name), used to reference the once occured alarm again, e.g. in case of updates (e.g. "temperature_sensor_loss") |
+| message-string    | Human readable short information about alarm reason (e.g. "Temperature sensor does not respond") |
+| severity-string   | Could be "CRITICAL", "MAJOR", "MINOR" or "WARNING" |
+| status-string     | Could be "ACTIVE" or "CLEARED"<br/>
+| time-string       | Timestamp indicating when the alarm (or the alarm-update) had occured (in [ISO 8601 format](http://www.w3.org/TR/NOTE-datetime#)) |
+
+# Data flow
+
+Figure below illustrates the data flow from Customer Application broker/thin-edge up to cloud. It shows especially the use of MQTT QOS=1, MQTT Retain and Mosquitto Persistence to achieve reliability alarm state-change transfer.
+
+![Sequence Diagram Update SW-list](images/alarm_dataflow.svg)
+
+**TODO:** Interface from Mapper to Cloud to be defined. Two options are possible:
+1) **JSONviaMQTT:** To have reliability use QOS=1. Open issue: Not yet completely confirmed that C8Y sends PUBACK when msg was processed, instead just on arrival.
+2) **HTTP REST:** HTTP response indicates if message was processed sucessfully. Anyway HTTP REST would increase complexity of implementation and increase data traffic.
+
+
+# Public MQTT-based alarm interface
+
+Similar to thin-edge'S measurement interface, alarm interface is based on MQTT topics.
+
+## Topic structure and payload
+
+Proposal to have "severity" and "ref-string" as topics:
+
+**thin-edge JSON (alarm) format:**
+```
+Topic:   tedge/alarms/<device-id>/<severity-string>/<ref-string>
+Payload: {
+  "message":  <message-string>,
+  "status":   <status-string>,  
+  "time":     <time-string>
+}
+```
+
+NOTE: The "device-id" indicates whether it is an alarm of the thin-edge device it-self (root-device) or a child-device. Use "root" for the thin-edge device or the child-device id for the according child-device.
+
+Benefit to have "severity" and "ref-string" as topics:
+Device-Site reactions to alarm could be easier relized.<br/>
+Examples:
+```
+# Listen to all major alarms on all devices
+tedge mqtt sub "tedge/alarms/+/major/+"
+
+# Listen to specific alarm on all devices
+# (e.g. a bunch of temperature sensors that have each an upper limit setpoint with alarm)
+tedge mqtt sub "tedge/alarms/+/+/upper-limit-exceeded"
+
+# And for sure all combinations.
+```
+
+
+Additionally an also supported "raw"-payload format will allow easy adoption for existing "MQTT publishers":
+
+**Raw payload format:**
+```
+Topic:   tedge/alarms/<device-id>/<severity-string>/<ref-string>
+Payload: <message-string>
+```
+In raw format for missing fields default values are used as below: 
+   - "status" := "ACTIVE", if payload is not empty; "CLEARED", if payload empty

```

Comment:
Could status be part of the topic?

Reply:
I missed this comment:
As we are utilizing the topics for the most of the handling why not push out the status there as well. It gives you even more flexibility with topics and you can only subscribe to ACTIVEation of the alarm.
I suppose for majority of cases all the most important information will be in the topic struct then and only if you really need to you can check detailed message or timestamp from payload.

Reply:
Technically yes, but this would be more involved for the alarm publisher. A state transition would imply the publication two messages: one to set the new state, another one to delete the message that was retained for the previous state.

Reply:
"CLEARED", if payload empty.  This makes sense but it might be tricky for the mapper to be aware that an alarm has been cleared while it was done. Indeed, a retained empty message is understood as *delete* the previous message. A connected client will be notified of the state change. But no messages are then published on reconnect.

Reply:
> [...] A connected client will be notified of the state change. But no messages are then published on reconnect.

That's true. But in combination with stored "sent message" for de-duplication mapper could detect also not yet sent empty message, in case of restart. I have added now a short description about that behavior into section about "raw payload format".